File: 프로그래머스/2/42746. 가장 큰 수/가장 큰 수.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool compare(int a, int b) {
    string c = to_string(a);
    string d = to_string(b);
    
    if (c[0] != d[0])  return c[0] > d[0];
    return c + d > d + c;
}


string solution(vector<int> numbers) {
    string answer = "";
    
    sort(numbers.begin(), numbers.end(), compare);
    
    bool check = false;
    
    for (int i = 0; i < numbers.size(); i++) {
        answer += to_string(numbers[i]);
        if (numbers[i] != 0) {
            check = true;
        }
    }
    
    if(check) return answer;
    else return "0";
}

File: 프로그래머스/2/42746. 가장 큰 수/가장 큰 수.java
import java.util.*;
import java.util.stream.Collectors;

class Solution {
    public String solution(int[] numbers) {
        String answer = "";
        
        List<String> slist = new ArrayList<>();
        
        for (int x : numbers) {
            String s = String.valueOf(x);
            slist.add(s);
        }
        
        // (30,3)인 경우 303과 330(b+a)을 비교해서 앞이 더 크므로 a와 b자리를 swap
        answer = slist.stream()
            .sorted((a, b) -> (b + a).compareTo(a + b))
            .collect(Collectors.joining());
        
        if (answer.charAt(0) == '0') return "0";
        
        return answer;
    }
}